daml/ContingentClaims/Math/Stochastic.daml

```diff
@@ -0,0 +1,252 @@
+-- places e.g. `Process`, where however not every choice is legal and will lead to
+-- a partial evaluator.
+data ExprF t x
+  = ConstF Decimal
+  | IdentF t
+  | ProcF { name : Text, process: Process t, filtration : t }
+  | SupF { lowerBound: t, tau: t, rv : x }
+  | AddF { lhs: x, rhs: x }
+  | NegF x
+  | MulF { lhs: x, rhs: x }
+  | PowF { lhs: x, rhs: x }
+  | I_F { lhs : x, rhs: x }
+  | E_F { rv : x, filtration: t }
+  deriving (Functor)
+
+-- | Represents an expression of t-adapted stochastic processes
+data Expr t
+  = Const Decimal
+  | Ident t
+  | Proc { name : Text, process : Process t, filtration : t }
+  | Sup { lowerBound: t, tau: t, rv : Expr t }
+  | Add (Expr t, Expr t)
+  | Neg (Expr t)
+  | Mul (Expr t, Expr t)
+  | Pow (Expr t, Expr t)
+  | I (Expr t , Expr t)
+  | E { rv : Expr t, filtration: t }
+  deriving (Show, Eq)
+
+instance Recursive (Expr t) (ExprF t) where
+  project (Const d) = ConstF d
+  project (Ident s) = IdentF s
+  project Proc{..} = ProcF with ..
+  project Sup{..} = SupF with ..
+  project (Add (x, x')) = AddF x x'
+  project (Neg x) = NegF x
+  project (Mul (x,x')) = MulF x x'
+  project (Pow (x, x')) = PowF x x'
+  project (I (x, x')) = I_F x x'
+  project E{..} = E_F with ..
+
+instance Corecursive (Expr t) (ExprF t) where
+  embed (ConstF d) = Const d
+  embed (IdentF s) = Ident s
+  embed ProcF{..} = Proc with ..
+  embed SupF{..} = Sup with ..
+  embed (AddF x x') = Add (x, x')
+  embed (NegF x) = Neg x
+  embed (MulF x x') = Mul (x, x')
+  embed (PowF x x') = Pow (x, x')
+  embed (I_F x x') = I (x, x')
+  embed E_F{..} = E with ..
+
+class IsIdentifier t where
+  -- | Produce a local identifier of type `t`, subindexed by `i`
+  localVar : Int -> t
+
+instance Foldable (ExprF t) where
+  foldMap f (ConstF _) = mempty
+  foldMap f (IdentF _) = mempty
+  foldMap f (ProcF {}) = mempty
+  foldMap f (SupF _ _ x) = f x
+  foldMap f (AddF x x') = f x <> f x'
+  foldMap f (NegF x) = f x
+  foldMap f (MulF x x') = f x <> f x'
+  foldMap f (PowF x x') = f x <> f x'
+  foldMap f (I_F x x') = f x <> f x'
+  foldMap f (E_F x t) = f x
+
+instance Traversable (ExprF t) where
+  sequence (ConstF d) = pure $ ConstF d
+  sequence (IdentF t) = pure $ IdentF t
+  sequence (ProcF {..}) = pure $ ProcF {..}
+  sequence (SupF t τ fa) = SupF t τ <$> fa
+  sequence (AddF fa fa') = AddF <$> fa <*> fa'
+  sequence (NegF fa) = NegF <$> fa
+  sequence (MulF fa fa') = MulF <$> fa <*> fa'
+  sequence (PowF fa fa') = PowF <$> fa <*> fa'
+  sequence (I_F fa fa') = I_F <$> fa <*> fa'
+  sequence (E_F fa t) = flip E_F t <$> fa
+
+{-# WARNING fapf "This is an experimental feature" #-}
+-- | Converts a `Claim` into the Fundamental Asset Pricing Formula. The ϵ
+-- expressions are defined as E1-E10 in the Eber/Peyton-Jones paper. If you
+-- squint you can almost see they correspond one-to-one to the formulae in our
+-- whitepaper.
+fapf : (Eq a, Show a, IsIdentifier t)
```

Comment:
I'm not thrilled by this `IsIdentifier` typeclass. The alternative is to put it as a parameter to this function `i -> t`, but I don't like this idea as it's an implementation detail rather than a business concern.